sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/SynchronousMetricStorage.java

```diff
@@ -154,30 +164,173 @@ public void recordDouble(double value, Attributes attributes, Context context) {
   }
 
   @Override
-  public MetricData collectAndReset(long startEpochNanos, long epochNanos) {
+  public MetricDescriptor getMetricDescriptor() {
+    return metricDescriptor;
+  }
+
+  /**
+   * This method leverages the `savedDeltas` and `reportHistory` to construct a final metric point
+   * to report for this synchronous instrument.
+   */
+  @GuardedBy("collectLock")
+  @Nullable
+  private MetricData buildMetricFor(
+      CollectionHandle collector, long startEpochNanos, long epochNanos) {
+    Map<Attributes, T> result = new HashMap<>();
+    // Next merge the delta w/ the last set of points.
+    for (DeltaAccumulation<T> point : savedDeltas) {
+      if (!point.wasReadBy(collector)) {
+        mergeInPlace(result, point.read(collector));
+      }
+    }
+
+    long lastEpochNanos;
+    // First pull the last cumulative value.
+    if (reportHistory.containsKey(collector)) {
+      LastReportedAccumulation<T> last = reportHistory.get(collector);
+      // Send the accumulated deltas in w/ previous accumulation to get final result.
+      if (aggregator.isStateful()) {
+        mergeInPlace(result, last.getAccumlation());
+      }
+      lastEpochNanos = last.getEpochNanos();
+    } else {
+      lastEpochNanos = startEpochNanos;
+    }
+
+    // Now write the aggregated value back, and generate final metric.
+    reportHistory.put(collector, new LastReportedAccumulation<>(result, epochNanos));
+    // Don't make a metric if we have no values.
+    if (result.isEmpty()) {
+      return null;
+    }
+    return aggregator.toMetricData(result, startEpochNanos, lastEpochNanos, epochNanos);
+  }
+
+  /** Merges accumulations from {@code toMerge} into {@code result}. */
+  private void mergeInPlace(Map<Attributes, T> result, Map<Attributes, T> toMerge) {
+    toMerge.forEach(
+        (k, v) -> {
+          if (result.containsKey(k)) {
+            result.put(k, aggregator.merge(result.get(k), v));
+          } else {
+            result.put(k, v);
+          }
+        });
+  }
+
+  /** Removes deltas once all collectors have pulled them. */
+  @GuardedBy("collectLock")
+  private void cleanup(Set<CollectionHandle> collectors) {
+    Iterator<DeltaAccumulation<T>> i = savedDeltas.iterator();
+    while (i.hasNext()) {
+      DeltaAccumulation<T> delta = i.next();
+      if (delta.wasReadyByAll(collectors)) {
+        i.remove();
+      }
+    }
+    // TODO: Do we allow different pipelines?
+    Iterator<CollectionHandle> c = reportHistory.keySet().iterator();
+    while (c.hasNext()) {
+      if (!collectors.contains(c.next())) {
+        c.remove();
+      }
+    }
+  }
```

Comment:
IDEA has some nifty suggestions for simplifying this code with the `Iterator.removeIf()` method.

Reply:
nice! done.